zebrad/src/components/mempool/storage.rs

```diff
@@ -139,4 +155,35 @@ impl Storage {
             .filter(|tx| self.rejected.contains_key(tx))
             .collect()
     }
+
+    /// Checks if the `tx` transaction conflicts with another transaction in the mempool.
+    ///
+    /// Two transactions conflict if they spent the same UTXO or if they reveal the same nullifier.
+    fn check_spend_conflicts(&self, tx: &UnminedTx) -> bool {
+        self.has_conflicts(tx, Transaction::inputs)
```

Comment:
Correctness: Identical `OutPoint`s spend the same UTXO, even if they have different `unlock_script`s or `sequence` numbers. So we need to compare the `PrevOut.outpoint`s here.

You might want to add a new method to `Transaction` that gets the spent `OutPoint`s.

If you do, you might want to refactor this existing code to use it as well.
(It's used to check the same consensus rules in the state.)
https://github.com/ZcashFoundation/zebra/blob/de589f6b3113af62c98e327edf969f2ef176a152/zebra-state/src/service/check/utxo.rs#L47-L52

Consensus Rules:

> each output of a particular transaction can only be used as an input once in the block chain.
> Any subsequent reference is a forbidden double spend - an attempt to spend the same satoshis twice.

https://developer.bitcoin.org/devguide/block_chain.html#introduction

Also, `Coinbase` inputs are only valid in blocks, not in the mempool.

This isn't explicitly documented, but it's implied in the Bitcoin docs, and implemented in `zcashd`:

> The first transaction in a block, called the coinbase transaction, must have exactly one input, called a coinbase. 

https://developer.bitcoin.org/reference/transactions.html#coinbase-input-the-input-of-the-first-transaction-in-a-block

Reply:
(I edited my comment to mention the `Input::outpoint` method.)

Reply:
Nice catch! I'll fix this tomorrow, thanks!